### PR TITLE
feat(recommend): Add trending types and methods

### PR DESCRIPTION
**Pull request by SophieManley03**
## What?
Prepare typing and methods for the new recommend model named `trending`. 

## Why? 
We want to release trending model in beta on March, 10th. In order to have a successful release we would like to update the recommend library so that use can use trending with it. 

## How?
[Here](https://algolia.atlassian.net/wiki/spaces/RECO/pages/4014473320/RFC+-+Recommend+-+Trending) are the specs for trending. It will give a better idea of what is implemented in the search.

We have 3 types of trending:
- trending items: Return trending items for a given index.
- trending items for a facet: Return trending items given a facetName and a facetValue.
- trending facet: Return trending facetValue given a facetName

We don't want to break the current implementation of recommend for our customers.

### CODE REVIEW

File: packages/recommend/src/__tests__/getTrendingItemsForFacet.test.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            },
+          ],
+        },
+        method: 'POST',
+        path: '1/indexes/*/recommendations',
+      },
+      {}
+    );
+  });
+});
```
Comment (SophieManley03) [inline packages/recommend/src/__tests__/getTrendingItemsForFacet.test.ts:48]: I add issue with Docker, and couldn't run the test suit on my side. But I guess it runs while building the PR?

**Review by jojva — APPROVED** [review]
LGTM :+1:

**Comment by SophieManley03** [comment]
Hi there, I made this first PR, I notice that the bundle size is bit too big, do you have any advice to make it smaller?

File: packages/recommend/src/types/RecommendationsQuery.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ export type RecommendationsQuery = {
   /**
    * The `objectID` of the item to get recommendations for.
    */
-  readonly objectID: string;
+  readonly objectID?: string;
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: it should be required, but omitted from the new items then probably, as how facetName is omitted for fbt

**Review by Haroenv — CHANGES_REQUESTED** [review]

File: packages/recommend/src/types/RecommendationsQuery.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (shortcuts) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: It is omitted actually, how come it's still needed?

**Comment by SophieManley03** [comment]
@Haroenv the problem if I do that is with getRecommendations . `getRecommendation` is used for every getX method.

File: packages/recommend/src/types/RecommendationsQuery.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SophieManley03) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: the problem if I do that is with getRecommendations . getRecommendation is used for every get method of Recommend and the objectID will be mandatory event for trends

[the remark this answers is not in the document]
Comment (francoischalifour) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: Can we still remodel the code so that types are accurate?

[the remark this answers is not in the document]
Comment (SophieManley03) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: I think so yes, if it does not break the current implementation. I'll do that. Thank you!

File: packages/recommend/src/methods/getTrendingItemsForFacet.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+type GetTrendingItemsForFacet = (
+  base: BaseRecommendClient
+) => WithRecommendMethods<BaseRecommendClient>['getTrendingItemsForFacet'];
+
+export const getTrendingItemsForFacet: GetTrendingItemsForFacet = base => {
+  return (queries: readonly TrendingItemsForFacetQuery[], requestOptions) => {
+    return getTrendingGlobalItems(base)(
+      queries.map(query => ({
+        ...query,
+      })),
```
Comment (Haroenv) [inline packages/recommend/src/methods/getTrendingItemsForFacet.ts:13]:
```suggestion
      queries,
```

this seems equivalent (you have a similar pattern in some other places)

**Review by Haroenv — COMMENTED** [review]
👌

File: packages/recommend/src/methods/getTrendingItemsForFacet.ts (hunk 3 above)
Comment (SophieManley03) [reply packages/recommend/src/methods/getTrendingItemsForFacet.ts:13]: Good remark, I forget to add the model attributes

File: packages/recommend/src/types/RecommendationsQuery.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SophieManley03) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: So after a third though, I wanted to try to make it optional here but required for RP and FBT using `required`. I did that to be able to use `getRecommendations` for all models.

[the remark this answers is not in the document]
Comment (PLNech) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: Put differently: was `getRecommendation` a hack while we defined the actual model names before GA, or is it a feature designed to allow using not-yet-public new models?
- If the former, then it might be time we drop this generic method (which brings subpar "common denominator" DX compared to specialized methods) and only implement Trending as a new, fully-typed method
- If the latter, we'll need to find a generic implementation that is least disruptive to current users.

[the remark this answers is not in the document]
Comment (francoischalifour) [reply packages/recommend/src/types/RecommendationsQuery.ts:18]: `getRecommendations()` is useful for the situation we're in right now: users can already use the trending model, without it being officially supported with a method in the client.

We didn't have a team to maintain this Recommend client initially so we provided it as an "escape hatch" if we're not reactive to support new models in the client.

`getRecommendations()` was just the very initial implementation that satisfied us at the time, feel free to change it as you need. And if it's not helpful for the new model, don't feel like you need to use it. (In the future, we can also deprecate it.)

File: packages/recommend/src/__tests__/getTrendingItems.test.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  test('builds the request', async () => {
+    const client = createMockedClient();
+
+    await client.getTrendingItems(
+      [
+        {
+          indexName: 'products',
+        },
+      ],
+      {}
```
Comment (shortcuts) [inline packages/recommend/src/__tests__/getTrendingItems.test.ts:22]: Is it also added to other tests? it shouldn't be required to pass the empty object

File: packages/recommend/src/types/FrequentlyBoughtTogetherQuery.ts [hunk 5]
```diff
@@ -2,5 +2,6 @@ import { RecommendationsQuery } from './RecommendationsQuery';
 
 export type FrequentlyBoughtTogetherQuery = Omit<
   RecommendationsQuery,
-  'model' | 'fallbackParameters'
->;
+  'model' | 'fallbackParameters' | 'facetName' | 'facetValue'
+> &
+  Required<Pick<RecommendationsQuery, 'objectID'>>;
```
Comment (shortcuts) [inline packages/recommend/src/types/FrequentlyBoughtTogetherQuery.ts:7]: I think utility types makes it less readable, I'd personally be fine with having both `RecommendationsQuery` and `TrendingsQuery` defined on their own in that case

**Review by shortcuts — DISMISSED** [review]
It looks good to me! 

Only one non-blocking comment if @francoischalifour have nothing agains it

File: packages/recommend/src/__tests__/getTrendingItems.test.ts (hunk 4 above)
Comment (shortcuts) [reply packages/recommend/src/__tests__/getTrendingItems.test.ts:22]: You can ignore my comment, thinking again, it allow us to ensure the body is empty so it's fine

**Review by PLNech — APPROVED** [review]
LGTM! Let's see how it goes once implemented in @algolia/recommend :upside_down_face:

**Review by Haroenv — APPROVED** [review]
looks good!

File: packages/recommend/src/types/RecommendModel.ts [hunk 6]
```diff
@@ -1 +1,2 @@
-export type RecommendModel = 'related-products' | 'bought-together';
+export type RecommendModel = 'related-products' | 'bought-together' | TrendingModel;
+export type TrendingModel = 'trending-items' | 'trending-facets';
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply packages/recommend/src/types/RecommendModel.ts:2]: Sorry D: https://github.com/algolia/algoliasearch-client-javascript/pull/1399

**Comment by SophieManley03** [comment]
Thank you all for the reviews 🙇‍♀️

File: packages/recommend/src/types/RecommendationsQuery.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jojva) [reply packages/recommend/src/types/RecommendationsQuery.ts]: I'll ask the opposite question then: what's the purpose of having a specialized method for each model rather than just using `getRecommendations()`? It's basically a multi-query endpoint, and I think it has its uses (like making a single call to retrieve both trending items and trending facets for example). While calling `getRelatedProducts()` as a multi-query is strange IMO.
